File: src/commands/backup.ts
const KIND = { CANCELLED: "cancelled", SELECTED: "selected", CREATED: "created", PREPARED: "prepared", RESTORED: "restored", ERROR: "error" } as const;
const ERROR_CODE = { STORAGE_UNAVAILABLE: "storage_unavailable", DESTINATION_EXISTS: "destination_exists", INVALID_BACKUP: "invalid_backup", UNSUPPORTED_SCHEMA: "unsupported_schema", CONFIRMATION_REQUIRED: "confirmation_required", TOKEN_INVALID: "token_invalid", TOKEN_EXPIRED: "token_expired", RESTORE_FAILED: "restore_failed", RECOVERY_FAILED: "recovery_failed", DATABASE_UNAVAILABLE: "database_unavailable" } as const;

export interface BackupSummary { path: string; created_at_unix_seconds: number; size_bytes: number; schema_version: number; }
export interface RestoreSummary { token: string; size_bytes: number; schema_version: number; }
export type PathSelection = { kind: typeof KIND.CANCELLED } | { kind: typeof KIND.SELECTED; path: string };
export type BackupResponse = { kind: typeof KIND.CREATED; summary: BackupSummary } | { kind: typeof KIND.PREPARED; summary: RestoreSummary } | { kind: typeof KIND.RESTORED } | BackupError;
export interface BackupError { kind: typeof KIND.ERROR; code: (typeof ERROR_CODE)[keyof typeof ERROR_CODE]; message: string; }
type Invoke = (command: string, payload: Record<string, unknown>) => Promise<unknown>;
type RecordValue = Record<string, unknown>;
const record = (value: unknown): value is RecordValue => typeof value === "object" && value !== null;
const error = (value: unknown): BackupError => ({ kind: KIND.ERROR, code: record(value) && typeof value.code === "string" && Object.values(ERROR_CODE).includes(value.code as BackupError["code"]) ? value.code as BackupError["code"] : ERROR_CODE.STORAGE_UNAVAILABLE, message: record(value) && typeof value.code === "string" && value.code === ERROR_CODE.DESTINATION_EXISTS ? "A backup already exists at that destination." : record(value) && typeof value.code === "string" && value.code === ERROR_CODE.INVALID_BACKUP ? "The selected backup is invalid." : record(value) && typeof value.code === "string" && value.code === ERROR_CODE.UNSUPPORTED_SCHEMA ? "The selected backup schema is unsupported." : record(value) && typeof value.code === "string" && value.code === ERROR_CODE.CONFIRMATION_REQUIRED ? "Restore confirmation is required." : record(value) && typeof value.code === "string" && value.code === ERROR_CODE.TOKEN_INVALID ? "The restore confirmation is invalid." : record(value) && typeof value.code === "string" && value.code === ERROR_CODE.TOKEN_EXPIRED ? "The restore confirmation has expired." : record(value) && typeof value.code === "string" && value.code === ERROR_CODE.RESTORE_FAILED ? "The restore could not be completed." : record(value) && typeof value.code === "string" && value.code === ERROR_CODE.DATABASE_UNAVAILABLE ? "The database is unavailable." : "Backup storage is unavailable." });
const selection = (value: unknown): PathSelection | BackupError => record(value) && value.kind === KIND.SELECTED && typeof value.path === "string" ? { kind: KIND.SELECTED, path: value.path } : record(value) && value.kind === KIND.CANCELLED ? { kind: KIND.CANCELLED } : error(value);
const response = (value: unknown): BackupResponse => record(value) && value.kind === KIND.CREATED && ["path"].every((key) => typeof value[key] === "string") && ["created_at_unix_seconds", "size_bytes", "schema_version"].every((key) => typeof value[key] === "number") ? { kind: KIND.CREATED, summary: { path: value.path as string, created_at_unix_seconds: value.created_at_unix_seconds as number, size_bytes: value.size_bytes as number, schema_version: value.schema_version as number } } : record(value) && value.kind === KIND.PREPARED && typeof value.token === "string" && ["size_bytes", "schema_version"].every((key) => typeof value[key] === "number") ? { kind: KIND.PREPARED, summary: { token: value.token, size_bytes: value.size_bytes as number, schema_version: value.schema_version as number } } : record(value) && value.kind === KIND.RESTORED ? { kind: KIND.RESTORED } : error(value);

export function createBackupCommands(invoke: Invoke) {
  const call = (command: string, payload: Record<string, unknown>) => invoke(command, payload).then(response).catch(error);
  return {
    chooseBackupDestination: () => invoke("choose_backup_destination_command", {}).then(selection).catch(error),
    chooseRestoreSource: () => invoke("choose_restore_source_command", {}).then(selection).catch(error),
    createBackup: (destination: string) => call("create_backup_command", { request: { destination } }),
    prepareRestore: (source: string) => call("prepare_restore_command", { request: { source } }),
    confirmRestore: (token: string) => call("confirm_restore_command", { request: { token, confirmed: true } }),
  };
}
const tauriInvoke: Invoke = async (command, payload) => (await import("@tauri-apps/api/core")).invoke(command, payload);
export const backupCommands = createBackupCommands(tauriInvoke);

File: src-tauri/src/commands/backup.rs
use std::{
    collections::HashMap,
    fs,
    path::{Path, PathBuf},
    time::{SystemTime, UNIX_EPOCH},
};

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::{
    infrastructure::{
        filesystem::{BackupStore, StorageError},
        sqlite::{create_snapshot, stage_and_validate},
    },
    DatabaseState,
};

pub const ALLOWED_COMMANDS: [&str; 5] = [
    "choose_backup_destination_command",
    "choose_restore_source_command",
    "create_backup_command",
    "prepare_restore_command",
    "confirm_restore_command",
];

const RESTORE_TOKEN_TTL_SECONDS: u64 = 900;

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CreateBackupRequest {
    pub destination: PathBuf,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PrepareRestoreRequest {
    pub source: PathBuf,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ConfirmRestoreRequest {
    pub token: String,
    pub confirmed: bool,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum PathSelection {
    Selected { path: PathBuf },
    Cancelled,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum BackupResponse {
    Created {
        path: PathBuf,
        created_at_unix_seconds: u64,
        size_bytes: u64,
        schema_version: i64,
    },
    Prepared {
        token: String,
        size_bytes: u64,
        schema_version: i64,
    },
    Restored,
    Error {
        code: &'static str,
        message: &'static str,
    },
}

impl BackupResponse {
    pub fn error(code: &'static str) -> Self {
        let message = match code {
            "database_unavailable" => "The database is unavailable.",
            "destination_exists" => "A backup already exists at that destination.",
            "invalid_backup" => "The selected backup is invalid.",
            "unsupported_schema" => "The selected backup schema is unsupported.",
            "confirmation_required" => "Restore confirmation is required.",
            "token_invalid" => "The restore confirmation is invalid.",
            "token_expired" => "The restore confirmation has expired.",
            "restore_failed" => "The restore could not be completed.",
            _ => "Backup storage is unavailable.",
        };
        Self::Error { code, message }
    }

    fn from_internal_code(code: &str) -> Self {
        match code {
            "database_unavailable" => Self::error("database_unavailable"),
            _ => Self::error("storage_unavailable"),
        }
    }
}

struct PendingRestore {
    stage: PathBuf,
    sha256: String,
    expires_at: u64,
}

pub struct BackupCommandState {
    root: PathBuf,
    pending: HashMap<String, PendingRestore>,
}

impl BackupCommandState {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self {
            root: root.into(),
            pending: HashMap::new(),
        }
    }
}

pub fn select_path(path: Option<PathBuf>) -> PathSelection {
    path.map_or(PathSelection::Cancelled, |path| PathSelection::Selected {
        path,
    })
}

#[cfg(feature = "desktop")]
pub async fn select_callback_path(
    open_picker: impl FnOnce(Box<dyn FnOnce(Option<PathBuf>) + Send>),
) -> PathSelection {
    let (sender, mut receiver) = tauri::async_runtime::channel(1);
    open_picker(Box::new(move |path| {
        let _ = sender.try_send(select_path(path));
    }));
    receiver.recv().await.unwrap_or(PathSelection::Cancelled)
}

pub fn create_backup(
    state: &DatabaseState,
    commands: &BackupCommandState,
    request: CreateBackupRequest,
) -> BackupResponse {
    let created_at_unix_seconds = now_seconds();
    let snapshot = commands.root.join("backup-restore/snapshots").join(format!(
        "{created_at_unix_seconds}-{}.sqlite3",
        uuid::Uuid::new_v4()
    ));
    let metadata = match state.with_read(|connection| {
        create_snapshot(connection, &snapshot).map_err(|_| "storage_unavailable".into())
    }) {
        Ok(metadata) => metadata,
        Err(code) => return BackupResponse::from_internal_code(&code),
    };
    let file_name = format!(
        "backup-{created_at_unix_seconds}-{}.sqlite3",
        uuid::Uuid::new_v4()
    );
    let published = BackupStore::new(&commands.root).publish_snapshot(
        &snapshot,
        &request.destination.join("backup-restore"),
        &file_name,
    );
    let _ = fs::remove_file(snapshot);
    match published {
        Ok(published) => BackupResponse::Created {
            path: published.path,
            created_at_unix_seconds,
            size_bytes: published.size_bytes,
            schema_version: metadata.schema_version,
        },
        Err(error) => BackupResponse::error(storage_error_code(error)),
    }
}

pub fn prepare_restore(
    state: &DatabaseState,
    commands: &mut BackupCommandState,
    request: PrepareRestoreRequest,
) -> BackupResponse {
    if let Err(code) = state.with_read(|_| Ok(())) {
        return BackupResponse::from_internal_code(&code);
    }
    let token = uuid::Uuid::new_v4().to_string();
    let stage = commands
        .root
        .join("backup-restore/staging")
        .join(format!("{token}.sqlite3"));
    let metadata = match stage_and_validate(&request.source, &stage) {
        Ok(metadata) => metadata,
        Err(crate::infrastructure::sqlite::BackupValidationError::InvalidBackup) => {
            return BackupResponse::error("invalid_backup")
        }
        Err(crate::infrastructure::sqlite::BackupValidationError::UnsupportedSchema) => {
            return BackupResponse::error("unsupported_schema")
        }
    };
    let Ok(size_bytes) = fs::metadata(&stage).map(|metadata| metadata.len()) else {
        return BackupResponse::error("storage_unavailable");
    };
    let Ok(sha256) = checksum(&stage) else {
        return BackupResponse::error("storage_unavailable");
    };
    commands.pending.insert(
        token.clone(),
        PendingRestore {
            stage,
            sha256,
            expires_at: now_seconds().saturating_add(RESTORE_TOKEN_TTL_SECONDS),
        },
    );
    BackupResponse::Prepared {
        token,
        size_bytes,
        schema_version: metadata.schema_version,
    }
}

pub fn confirm_restore(
    state: &DatabaseState,
    commands: &mut BackupCommandState,
    request: ConfirmRestoreRequest,
) -> BackupResponse {
    let Some(pending) = commands.pending.get(&request.token) else {
        return BackupResponse::error("token_invalid");
    };
    if now_seconds() >= pending.expires_at {
        commands.pending.remove(&request.token);
        return BackupResponse::error("token_expired");
    }
    if !request.confirmed {
        return BackupResponse::error("confirmation_required");
    }
    let Some(pending) = commands.pending.remove(&request.token) else {
        return BackupResponse::error("token_invalid");
    };
    if !matches!(checksum(&pending.stage), Ok(checksum) if checksum == pending.sha256) {
        return BackupResponse::error("invalid_backup");
    }
    match state.install_validated_stage(&pending.stage, &BackupStore::new(&commands.root)) {
        Ok(()) => BackupResponse::Restored,
        Err(code) if code == "database_unavailable" => {
            BackupResponse::error("database_unavailable")
        }
        Err(_) => BackupResponse::error("restore_failed"),
    }
}

fn storage_error_code(error: StorageError) -> &'static str {
    match error {
        StorageError::DestinationExists => "destination_exists",
        StorageError::SelectionCancelled | StorageError::StorageUnavailable => {
            "storage_unavailable"
        }
    }
}

fn checksum(path: &Path) -> std::io::Result<String> {
    use std::io::Read;

    let mut file = fs::File::open(path)?;
    let mut digest = Sha256::new();
    let mut buffer = [0; 8192];
    loop {
        let count = file.read(&mut buffer)?;
        if count == 0 {
            return Ok(format!("{:x}", digest.finalize()));
        }
        digest.update(&buffer[..count]);
    }
}

fn now_seconds() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_or(0, |duration| duration.as_secs())
}
